cpp/include/cudf/fixed_point/fixed_point.hpp

```diff
@@ -0,0 +1,289 @@
+    auto const scale = lhs._scale > rhs._scale ? lhs._scale : rhs._scale;
+
+    #if defined(__CUDACC_DEBUG__)
+
+    if (addition_overflow<Rep1>(lhsv, rhsv))
+        CUDF_FAIL("fixed_point overflow of underlying represenation type " + print_rep<Rep1>());
+
+    #endif
+
+    return fixed_point<Rep1, Rad1>{scaled_integer<Rep1>(lhsv + rhsv, scale)};
+}
+
+// MINUS Operation
+template<typename Rep1, Radix Rad1,
+         typename Rep2, Radix Rad2>
+fixed_point<Rep1, Rad1> operator-(fixed_point<Rep1, Rad1> const& lhs,
+                                  fixed_point<Rep2, Rad2> const& rhs) {
+
+    static_assert(std::is_same<Rep1, Rep2>::value, "Represenation types should be the same");
+    static_assert(Rad1 == Rad2,                    "Radix types should be the same");
+
+    auto const rhsv  = lhs._scale > rhs._scale ? detail::shift<Rad1>(rhs._value, scale_type{lhs._scale - rhs._scale}) : rhs._value;
+    auto const lhsv  = lhs._scale < rhs._scale ? detail::shift<Rad1>(lhs._value, scale_type{rhs._scale - lhs._scale}) : lhs._value;
+    auto const scale = lhs._scale > rhs._scale ? lhs._scale : rhs._scale;
+
+    #if defined(__CUDACC_DEBUG__)
+
+    if (subtraction_overflow<Rep1>(lhsv, rhsv))
+        CUDF_FAIL("fixed_point overflow of underlying represenation type " + print_rep<Rep1>());
+
+    #endif
+
+    return fixed_point<Rep1, Rad1>{scaled_integer<Rep1>(lhsv - rhsv, scale)};
+}
+
+// MULTIPLIES Operation
+template<typename Rep1, Radix Rad1,
+         typename Rep2, Radix Rad2>
+fixed_point<Rep1, Rad1> operator*(fixed_point<Rep1, Rad1> const& lhs,
+                                  fixed_point<Rep2, Rad2> const& rhs) {
+
+    static_assert(std::is_same<Rep1, Rep2>::value, "Represenation types should be the same");
+    static_assert(Rad1 == Rad2,                    "Radix types should be the same");
+
+    #if defined(__CUDACC_DEBUG__)
+
+    if (multiplication_overflow<Rep1>(lhs._value, rhs._value))
+        CUDF_FAIL("fixed_point overflow of underlying represenation type " + print_rep<Rep1>());
+
+    #endif
+
+    return fixed_point<Rep1, Rad1>{scaled_integer<Rep1>(lhs._value * rhs._value, scale_type{lhs._scale + rhs._scale})};
+}
+
+// DIVISION Operation
+template<typename Rep1, Radix Rad1,
+         typename Rep2, Radix Rad2>
+fixed_point<Rep1, Rad1> operator/(fixed_point<Rep1, Rad1> const& lhs,
+                                  fixed_point<Rep2, Rad2> const& rhs) {
+
+    static_assert(std::is_same<Rep1, Rep2>::value, "Represenation types should be the same");
+    static_assert(Rad1 == Rad2,                    "Radix types should be the same");
+
+    #if defined(__CUDACC_DEBUG__)
+
+    if (division_overflow<Rep1>(lhs._value, rhs._value))
+        CUDF_FAIL("fixed_point overflow of underlying represenation type " + print_rep<Rep1>());
+
+    #endif
+
+    return fixed_point<Rep1, Rad1>{scaled_integer<Rep1>(lhs._value / rhs._value, scale_type{lhs._scale - rhs._scale})};
```

Comment:
I'm wondering if it would be a good idea to shift `lhs._value` before division (assuming it's positive) to help with cases where you might underflow. For example `decimal32{60, scale_type{1}} / decimal32{10, scale_type{0}}` would return `decimal32{6/10, scale_type{1-0}}` which underflows to 0, but if you shift first you'd get `decimal32{60/10, scale_type{0-0}}` and save the 6.

Reply:
So this is a really good point and something it took me a while to wrap my head around. The correct answer is actual zero (see the CNL example [here](https://godbolt.org/z/BosGk8)). I have added a unit test that hopefully makes it a bit clearer.
```cpp
TEST_F(FixedPointTest, Decimal32TrickyDivision) {

    decimal32 ONE_1  {1,  scale_type{1}};
    decimal32 SIX_1  {6,  scale_type{1}};
    decimal32 TEN_0  {10, scale_type{0}};
    decimal32 TEN_1  {10, scale_type{0}};
    decimal32 SIXTY_1{60, scale_type{1}};

    decimal32 ZERO = ONE_1;

    EXPECT_EQ(ONE_1.get(),    0); // 1 / 10 = 0
    EXPECT_EQ(SIX_1.get(),    0); // 6 / 10 = 0
    EXPECT_EQ(TEN_0.get(),   10);
    EXPECT_EQ(SIXTY_1.get(), 60);

    EXPECT_EQ(SIXTY_1 / TEN_0, ZERO);
    EXPECT_EQ(SIXTY_1 / TEN_1, SIX_1);

}
```
With a positive scale, the values you store internally are divided by the `Radix ^ scale`, so for `scale = 1` and `Radix = 10`, we are essentially storing values **divided by ten**, which means that we can only actually represent **multiples of ten**. Note that if we divide by ten with `Radix = 1`, the resulting value has `Radix = 0` so we do end up with the expected answer of 6.

Does that make sense? I took me a few hours of playing around with CNL and many examples to wrap my head around this.

Reply:
Shouldn't TEN_1 be {10, scale_type{*1*}}?

Reply:
The Wikipedia article you linked above says this in the Operations section:
"One can obtain a more accurate result by first converting the dividend to a more precise type: in the same example, converting 3456 scaled by 1/100 (34.56) to 3,456,000 scaled by 1/100000, before dividing by 1234 scaled by 1/1000 (1.234), would yield 3456000÷1234 = 2801 (rounded) with scaling factor (1/100000)/(1/1000) = 1/100, that is 28.01 (instead of 30)."

Reply:
yes, you are right! fixed it (and pushed a small commit). I was still comparing zero to zero. This compares 6 to 6 now.

Reply:
"Converting to a more precise type" is the shifting before division that I was suggesting. I'm not sure if this is the right thing to do though, or if it's more correct to just allow the underflow.

Reply:
Let me add a unit test for that example and compare with CNL as well. I tested with doing the conversion first and it seemed to yield the same answers, but this looks like you might be right.

Reply:
`CNL` gives 20 as the answer: https://godbolt.org/z/7MzAa4
```cpp
auto a = cnl::fixed_point<int, -2, 10>{34.56};
auto b = cnl::fixed_point<int, -3, 10>{1.234};

std::cout << a / b << '\n'; // 20
```
and so does `cudf::fp::fixed_point`
```cpp
decimal32 A{34.56,  scale_type{-2}};
decimal32 B{1.234,  scale_type{-3}};

EXPECT_EQ((A / B).get(), 20);
```
The resulting value is effectively `value = 34.56 / 1.234 = 28.0064829822` but then because our resulting `scale = 1 = (-2 - (-3))`, we divide by `10^1` and the integer division leaves us with 20. So even though there might be a more accurate technique, based on how we have designed ours, I think we want to keep this behaviour.

Reply:
Ok, I think matching the behavior of CNL is a fair choice. If more precision is needed, the user can choose to scale up the dividend before division.

Reply:
I think we have different tradeoffs than CNL. In CNL the scale factor is compile-time information. So presumably the user has more information at compile time about whether more precision is needed. In libcudf the scale factor is run-time information. In a big data application, at run time one might not be able to reason about the required precision for any given operation. So it might be better to err on the side of achieving better accuracy where possible. The example @trevorsm7 quotes above is illustrative. If you were to assume the fixed-point decimal values in the example represent bank balances (34.56) and fractional days (1.234), and dividing them, one might be surprised to get an error of over $10 per day (43% relative error) in the result.  I guess as Trevor points out one can scale the balance column first. I just wonder if our users will understand this well enough to do so, and if so, to choose the right scale factor. Perhaps a helper function is needed?